homeassistant/components/elevenlabs/tts.py

```diff
@@ -0,0 +1,124 @@
+"""Support for the ElevenLabs text-to-speech service."""
+
+from __future__ import annotations
+
+from functools import cached_property
+import logging
+from typing import Any
+
+from elevenlabs.client import AsyncElevenLabs
+from elevenlabs.core import ApiError
+from elevenlabs.types import Model, Voice
+
+from homeassistant.components import tts
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import CONF_MODEL, CONF_VOICE, DEFAULT_MODEL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up ElevenLabs tts platform via config entry."""
+    model_id = config_entry.options.get(CONF_MODEL, config_entry.data.get(CONF_MODEL))
+    default_voice_id = config_entry.options.get(
+        CONF_VOICE, config_entry.data.get(CONF_VOICE)
+    )
+    client = AsyncElevenLabs(api_key=config_entry.data[CONF_API_KEY])
+
+    # Load model and voices here in async context
+    model: Model | None = None
+    models = await client.models.get_all()
+    model_id = model_id if model_id is not None else DEFAULT_MODEL
+    for maybe_model in models:
+        if maybe_model.model_id == model_id:
+            model = maybe_model
+            break
+
+    if (model is None) or (not model.languages):
+        raise ValueError(f"Failed to load model for id: {model_id}")
+
+    voices = (await client.voices.get_all()).voices
+
+    async_add_entities(
+        [ElevenLabsTTSEntity(config_entry, client, model, voices, default_voice_id)]
+    )
+
+
+class ElevenLabsTTSEntity(tts.TextToSpeechEntity):
+    """The ElevenLabs API entity."""
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        client: AsyncElevenLabs,
+        model: Model,
+        voices: list[Voice],
+        default_voice_id: str | None,
+    ) -> None:
+        """Init ElevenLabs TTS service."""
+        self._client = client
+        self._model = model
+        self._default_voice_id = default_voice_id
+        self._voices = sorted(
+            (tts.Voice(v.voice_id, v.name) for v in voices if v.name),
+            key=lambda v: v.name,
+        )
+        # Default voice first
+        voice_indices = [
+            idx for idx, v in enumerate(self._voices) if v.voice_id == default_voice_id
+        ]
+        if voice_indices:
+            self._voices.insert(0, self._voices.pop(voice_indices[0]))
+        self._attr_name = f"ElevenLabs {self._model.model_id}"
```

Comment:
Probably easiest to default to `config_entry.title`, so if user were to change that, it will trickle down.

Reply:
Done